File: 2017/day05-2.hs
solution :: String -> Int
solution = play 0 0 . map read . lines

play :: Int -> Int -> [Int] -> Int
play step index jumps 
    | index < 0 || index >= length jumps = step
    | otherwise = play (step +1) (index + (!!index) jumps) (take index jumps ++ [(!!index) jumps + (if' ((!!index) jumps < 3) (1) (-1))] ++ drop (index+1) jumps) 

if' :: Bool -> a -> a -> a
if' True a _ = a
if' False _ a = a

main :: IO ()
main = do readFile "day05.dat" >>= print . solution

File: 2017/day05-2.ts
function solution(content: string) {
  return play(content.split("\n").map(Number));
}

function play(jumps: number[]) {
  let step = 0;
  let index = 0;
  const copy = jumps.slice();
  while (0 <= index && index < copy.length) {
    const offset = copy[index];
    copy[index] += offset < 3 ? 1 : -1;
    index += offset;
    step += 1;
  }
  return step;
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
